src/pre-index-views.js

```diff
@@ -0,0 +1,129 @@
+const utils = require('./utils');
+const viewIndexerProgress = require('./view-indexer-progress');
+
+const buildsUrl = 'https://staging.dev.medicmobile.org/_couch/builds_4';
+const DDOC_DB_RE = /^ddocs\/(.+)\.json$/;
+const TIMEOUT_ERROR = 'timeout';
+
+const decodeAttachmentData = (data) => {
+  try {
+    const buffer = Buffer.from(data, 'base64');
+    return JSON.parse(buffer.toString('utf-8'));
+  } catch (err) {
+    console.error('Error while decoding attachment data');
+    throw err;
+  }
+};
+
+const getDDocs = async (version) => {
+  const buildUrl = `${buildsUrl}/medic:medic:${version}?attachments=true`;
+  let build;
+  try {
+    build = await utils.request({ url: buildUrl });
+  } catch (err) {
+    if (err.status === 404) {
+      err.message = `Build for version ${version} was not found.`;
+    } else {
+      err.message = `An unexpected error occurred when getting build for version ${version}`;
+    }
+    throw err;
+  }
+
+  const ddocs = Object.keys(build._attachments).filter(attachment => DDOC_DB_RE.test(attachment));
+  const ddocsForDb = Object.assign(...ddocs.map(ddoc => {
+    const match = ddoc.match(DDOC_DB_RE);
+    let dbName = match && match[1];
+
+    const attachment = decodeAttachmentData(build._attachments[ddoc].data);
+    if (!attachment || !attachment.docs || !attachment.docs.length) {
+      return {};
+    }
+
+    dbName = dbName === 'medic' ? dbName : `medic-${dbName}`;
+    return { [dbName]: attachment.docs };
+  }));
+
+  return ddocsForDb;
+};
+
+const dbExists = async (dbName) => {
+  try {
+    const url = utils.getUrl(dbName);
+    await utils.request({ url });
+    return true;
+  } catch (err) {
+    if (err.status === 404) {
+      return false;
+    }
+    throw err;
+  }
+};
+
+const getStagedDdocs = async (ddocsForDb) => {
+  const stagedDdocs = {};
+  for (const [dbName, ddocs] of Object.entries(ddocsForDb)) {
+    if (!await dbExists(dbName)) {
+      continue;
+    }
+
+    stagedDdocs[dbName] = ddocs.map(ddoc => {
+      ddoc._id = ddoc._id.replace('_design/', '_design/:staged:');
+      ddoc.version = `preindex-${ddoc.version}`;
+
+      return ddoc;
+    });
+  }
+
+  return stagedDdocs;
+};
+
+const saveStagedDdocs = async (stagedDdocs) => {
+  for (const [dbName, ddocs] of Object.entries(stagedDdocs)) {
+    const url = utils.getUrl(`${dbName}/_bulk_docs`);
+    await utils.request({ url, body: { docs: ddocs }, method: 'POST' });
+  }
+};
+
+const indexView = async (dbName, ddocId, viewName) => {
+  do {
+    try {
+      const url = utils.getUrl(`/${dbName}/${ddocId}/_view/${viewName}`, false, `limit=0`);
+      return await utils.request({ url });
+    } catch (requestError) {
+      if (!requestError || !requestError.response || !TIMEOUT_ERROR.includes(await requestError.response.error)) {
+        throw requestError;
+      }
```

Comment:
It's not obvious how this loop continues. Maybe pull out a variable for `timeoutError` to make this self documenting?

Also, do you really need to `await` the `error` access??

```suggestion
      const isTimeout = requestError && requestError.response && TIMEOUT_ERROR.includes(await requestError.response.error);
      if (isTimeout) {
        continue; // views still indexing - try again
      }
      throw requestError;
```

Reply:
> Also, do you really need to await the error access??

Without the error title, and just the status code, it's sometimes hard to make out what the error is. For example, a CouchDb timeout error is a 500, you only know it's a timeout if you inspect the body. 
I found it really useful with request-promise-native handed you the body of the errored response.

Reply:
On second thought, it seems like I had changed the error handling code in request to wait for the error content ! thanks for flagging this!